README.md
[![GitHub release](https://img.shields.io/github/release/SidebarJS/sidebarjs.svg)](https://github.com/SidebarJS/sidebarjs/releases)
[![npm](https://img.shields.io/npm/v/sidebarjs.svg)](https://www.npmjs.com/package/sidebarjs)
[![npm](https://img.shields.io/npm/dt/sidebarjs.svg)](https://www.npmjs.com/package/sidebarjs)

# SidebarJS
Create mobile sidebar/sidenav experiance in pure javascript.

```ssh
npm install sidebarjs --save
```

## Demo
*Open the demo on your device and try the touch gestures!*

* [StackBlitz](https://stackblitz.com/edit/sidebarjs)
* [RawGit](https://rawgit.com/SidebarJS/sidebarjs/master/demo/index.html)

## Libraries
* [Angular | ng-sidebarjs](https://github.com/SidebarJS/ng-sidebarjs)
* [AngularJS | angular-sidebarjs](https://github.com/SidebarJS/angular-sidebarjs)

## Import

Typescript / ES6
```js
import {SidebarElement, SidebarService} from 'sidebarjs';
import 'sidebarjs/lib/sidebarjs.css';
```

Classic
```html
<script src="your/path/sidebarjs.js"></script>
<link rel="stylesheet" href="your/path/sidebarjs.css">
```

## Options
```typescript
const sidebarjs = new SidebarJS.SidebarElement({
    // Sidebar DOM element
    component?: <HTMLElement>sidebarjs,
    
    // Minimum swipe in px required to trigger listener: open
    documentMinSwipeX?: 10,
    
    // Range in px where document is listening for gesture: open
    documentSwipeRange?: 40,
    
    // Open and close sidebar with swipe gestures
    nativeSwipe?: true,
    
    // Enable/Disable open on swipe
    nativeSwipeOpen?: true,
    
    // Sidebar position, accepted values: left|right
    position?: 'left',
    
    // Backdrop opacity on sidebar open
    backdropOpacity?: 0.3,
    
    // Show sidebar on screen > 1024px
    responsive?: false,
    
    // Page content if sidebar has option responsive
    mainContent?: <HTMLElement>
    
    // Function called after sidebar is open
    onOpen?: () => void,
    
    // Function called after sidebar is close
    onClose?: () => void,
    
    // Function called when sidebar change visibility
    onChangeVisibility?: (changes: { isVisible: boolean }) => void,
})
```

## Instance methods
### .open(): void
> Open sidebar
```typescript
const sidebarjs = new SidebarJS.SidebarElement();
sidebarjs.open();
```

### .close(): void
> Close sidebar
```typescript
const sidebarjs = new SidebarJS.SidebarElement();
sidebarjs.open();
setTimeout(() => {
  sidebarjs.close();
}, 3000);
```

### .toggle(): void
> Toggle sidebar
```typescript
const sidebarjs = new SidebarJS.SidebarElement();
sidebarjs.toggle();
```

### .isVisible(): boolean
> Check if sidebar is visible
```typescript
const sidebarjs = new SidebarJS.SidebarElement();
sidebarjs.isVisible(); // false
sidebarjs.open();
sidebarjs.isVisible(); // true
```

### .destroy(): void
> Destroy sidebar and all listeners
```typescript
const sidebarjs = new SidebarJS.SidebarElement();
sidebarjs.destroy();
```

### .setPosition(position: 'left'|'right'): void
> Set sidebar position
```typescript
const sidebarjs = new SidebarJS.SidebarElement();
sidebarjs.setPosition('right'); // sidebar move to the right side
```

### .setSwipeGestures(value: boolean): void
> Check nativeSwipe and nativeSwipeOpen props and enable/disable gestures only if prop is true
```typescript
const sidebarjs = new SidebarJS.SidebarElement({
  nativeSwipe: true,
  nativeSwipeOpen: false,
});
sidebarjs.setSwipeGestures(false); // disable only nativeSwipe listeners
sidebarjs.nativeSwipeOpen = true;
sidebarjs.setSwipeGestures(true); // enable nativeSwipe and nativeSwipeOpen listeners
```

## Single Sidebar
```html
<head>

  <link rel="stylesheet" href="sidebarjs.min.css">

</head>
<body>

  <div sidebarjs-toggle>Open/Close</div>

  <div sidebarjs>
    <nav>
      <a href="link">Home</a>
      <a href="link">About</a>
      <a href="link">Contacts</a>
    </nav>
  </div>

  <script src="sidebarjs.min.js"></script>
  <script>
  // Init SidebarJS
  var sidebarjs = new SidebarJS.SidebarElement();
  </script>

</body>
```

## Multiple Sidebars
```html
<head>

  <link rel="stylesheet" href="sidebarjs.min.css">

</head>
<body>

  <div sidebarjs-toggle="leftSidebarName">Open/Close Left Sidebar</div>
  <div sidebarjs-toggle="rightSidebarName">Open/Close Right Sidebar</div>

  <div sidebarjs="leftSidebarName">
    <nav>
      <a href="link">My</a>
      <a href="link">Left</a>
      <a href="link">Content</a>
    </nav>
  </div>

  <div sidebarjs="rightSidebarName">
    <nav>
      <a href="link">My</a>
      <a href="link">Right</a>
      <a href="link">Content</a>
    </nav>
  </div>

  <script src="sidebarjs.min.js"></script>
  <script>
  var leftSidebarjs = new SidebarJS.SidebarElement({
    component: document.querySelector('[sidebarjs="leftSidebarName"]'),
  });
  
  var rightSidebarjs = new SidebarJS.SidebarElement({
    component: document.querySelector('[sidebarjs="rightSidebarName"]'),
    position: 'right',
  });
  </script>

</body>
```

## Responsive Sidebar
```html
<head>

  <link rel="stylesheet" href="sidebarjs.min.css">

</head>
<body>

  <div sidebarjs>
    <nav>
      <a href="link">Home</a>
      <a href="link">About</a>
      <a href="link">Contacts</a>
    </nav>
  </div>
  
  <div sidebarjs-content>
    your content
    <div sidebarjs-toggle>Open/Close</div>
  </div>

  <script src="sidebarjs.min.js"></script>
  <script>
  // Init SidebarJS
  var sidebarjs = new SidebarJS.SidebarElement({responsive: true});
  </script>

</body>
```

## Trigger onOpen/onClose/onChangeVisibility
```html
<head>

  <link rel="stylesheet" href="sidebarjs.min.css">

</head>
<body>

  <div sidebarjs-toggle>Open/Close</div>

  <div sidebarjs>
    <nav>
      <a href="link">Home</a>
      <a href="link">About</a>
      <a href="link">Contacts</a>
    </nav>
  </div>

  <script src="sidebarjs.min.js"></script>
  <script>
  // Init SidebarJS
  var sidebarjs = new SidebarJS.SidebarElement({
    onOpen: function() {
      console.log('sidebarjs is open');
    },
    onClose: function() {
      console.log('sidebarjs is close');
    },
    onChangeVisibility: function(changes) {
      console.log('sidebarjs is visible?', changes.isVisible);
    }
  });
  </script>

</body>
```

## Destroy Sidebar
```html
<head>

  <link rel="stylesheet" href="sidebarjs.min.css">

</head>
<body>

  <div sidebarjs-toggle>Open/Close</div>

  <div sidebarjs>
    <nav>
      <a href="link">Home</a>
      <a href="link">About</a>
      <a href="link">Contacts</a>
    </nav>
  </div>

  <script src="sidebarjs.min.js"></script>
  <script>
  // Init SidebarJS
  var sidebarjs = new SidebarJS.SidebarElement();
  
  // After 4 seconds
  setTimeout(function() {
    
    // Destroy sidebarjs
    sidebarjs.destroy();
    
  }, 4000);
  </script>

</body>
```

CHANGELOG.md
# [10.0.0](https://github.com/SidebarJS/sidebarjs/compare/9.0.0...10.0.0) (2022-12-29)


### Features

* update all deps and change build process ([10ca7f0](https://github.com/SidebarJS/sidebarjs/commit/10ca7f0603d17f9b351c8e95407c69ed9c14c898))



# [9.0.0](https://github.com/SidebarJS/sidebarjs/compare/8.0.0...9.0.0) (2021-07-20)


### Features

* update all dev deps ([08ed445](https://github.com/SidebarJS/sidebarjs/commit/08ed44555f564b920a3c1b204ee722aab6308a7f))



# [8.0.0](https://github.com/SidebarJS/sidebarjs/compare/7.0.0...8.0.0) (2021-05-14)


### Features

* update deps + export SidebarPosition as value ([30cb894](https://github.com/SidebarJS/sidebarjs/commit/30cb894bf5bb7b651ba56a0b7353817066b58cfa))



# [7.0.0](https://github.com/SidebarJS/sidebarjs/compare/5.4.0...7.0.0) (2020-12-07)


### Features

* add transition-start|end classes to detect open/close actions via DOM | close [#14](https://github.com/SidebarJS/sidebarjs/issues/14) ([0b3a33c](https://github.com/SidebarJS/sidebarjs/commit/0b3a33c323670ba7e70f6b31c3f155afde9cf9c9))
* **eslint:** move from tslint to eslint ([082c579](https://github.com/SidebarJS/sidebarjs/commit/082c579ab8bffd5373f7796aad47e205ca945e13))
* **npm:** update all dependencies ([25ad50e](https://github.com/SidebarJS/sidebarjs/commit/25ad50e280ca331f47fcec10513d71a8123cc0dd))


### Performance Improvements

* bundle size ([2d916ba](https://github.com/SidebarJS/sidebarjs/commit/2d916babf140ecc0b5390a8b122e3656e79a4d91))



# [5.4.0](https://github.com/SidebarJS/sidebarjs/compare/5.3.0...5.4.0) (2018-11-30)


### Features

* **SidebarElement:** setSwipeGestures [#13](https://github.com/SidebarJS/sidebarjs/issues/13) ([2f37fab](https://github.com/SidebarJS/sidebarjs/commit/2f37fab7f650a7edc33f8837f3c4c23bb4e4109b))



# [5.3.0](https://github.com/SidebarJS/sidebarjs/compare/5.2.1...5.3.0) (2018-09-03)



## [5.2.1](https://github.com/SidebarJS/sidebarjs/compare/5.2.0...5.2.1) (2018-05-31)


### Bug Fixes

* **SidebarElement:** prevent close sidebar on opposite swipe gesture | [#9](https://github.com/SidebarJS/sidebarjs/issues/9) ([4a7a1d1](https://github.com/SidebarJS/sidebarjs/commit/4a7a1d115b5ddbd81e6e3a371f0fca52cdd968d2))


### Performance Improvements

* **SidebarElement:** use CSSOM when available | [#11](https://github.com/SidebarJS/sidebarjs/issues/11) ([6efb7e5](https://github.com/SidebarJS/sidebarjs/commit/6efb7e5acee839332d59e7ed7de0d1c0cb697b53))



# [5.2.0](https://github.com/SidebarJS/sidebarjs/compare/5.1.0...5.2.0) (2018-04-14)


### Features

* **SidebarElement:** responsive option ([3dc8d15](https://github.com/SidebarJS/sidebarjs/commit/3dc8d15f7a546b939410fb6e93e73e6d5a8178f4))



# [5.1.0](https://github.com/SidebarJS/sidebarjs/compare/5.0.0...5.1.0) (2018-03-07)


### Bug Fixes

* **SidebarElement:** correct trigger documentMinSwipeX on open ([a820765](https://github.com/SidebarJS/sidebarjs/commit/a8207650a92cfa16f6ff02e6f37f0e332ca90266))



# [5.0.0](https://github.com/SidebarJS/sidebarjs/compare/4.1.0...5.0.0) (2018-02-11)


### Bug Fixes

* **sidebarElement:** destroy method ([5bcd2b4](https://github.com/SidebarJS/sidebarjs/commit/5bcd2b4a1a8d6164ac2a7f42c618ef86182f9486))
* **sidebarElement:** destroy method ([a207f38](https://github.com/SidebarJS/sidebarjs/commit/a207f38b043b1b37ae67daa63ceb3ea32830b58e))
* **SidebarElement:** remove unnecessary will-change prop ([e4b2b62](https://github.com/SidebarJS/sidebarjs/commit/e4b2b62440b4f52fe3b0b07fb8311420d7c6ca60))
* **sidebarElement:** transclude ([75998f0](https://github.com/SidebarJS/sidebarjs/commit/75998f00a62d5cb760c5696c5be1f4be2ad1c175))


### Features

* **SidebarElement:** emit onOpen/onClose/onChangeVisibility when sidebar transition end ([4ffd5b9](https://github.com/SidebarJS/sidebarjs/commit/4ffd5b9a3714a7eae5dbd6bcc5d449e64c582c57))
* **SidebarElement:** implemented onOpen/onClose/onChangeVisibility in dist ([866fbdc](https://github.com/SidebarJS/sidebarjs/commit/866fbdcd6d0b7ca689a4bd448231864f1efbacfc))


### Performance Improvements

* **SidebarElement:** __onTouchMove / __onSwipeOpenMove ([7ff77e2](https://github.com/SidebarJS/sidebarjs/commit/7ff77e282daaefe9977252063f63dbf2a4eceb76))
* **sidebarElement:** component will-change ([78030ea](https://github.com/SidebarJS/sidebarjs/commit/78030ea24bd84f1206ea88b01ed3f1fee045120f))
* **sidebarElement:** drop support IE9 ([19f9984](https://github.com/SidebarJS/sidebarjs/commit/19f998477f71cde194824e5ce4dc4c050b93edb8))
* **sidebarElement:** no more detached DOM nodes ([28061b2](https://github.com/SidebarJS/sidebarjs/commit/28061b2ac8c02a741d8cffdec2719c3a1f59cbf8))



# [4.1.0](https://github.com/SidebarJS/sidebarjs/compare/4.0.0...4.1.0) (2018-02-05)


### Performance Improvements

* **sidebarElement:** passive listeners + will change ([ded4d67](https://github.com/SidebarJS/sidebarjs/commit/ded4d67aa222ad4f99868260ded3259750462699))



# [4.0.0](https://github.com/SidebarJS/sidebarjs/compare/3.2.0...4.0.0) (2018-01-13)


### Features

* **idebarElement:** configurable background opacity ([00a0673](https://github.com/SidebarJS/sidebarjs/commit/00a0673bd77dff2a608c98db30bcb1e794b0b121))


### Performance Improvements

* **css:** transition: all > transform ([cf81e1a](https://github.com/SidebarJS/sidebarjs/commit/cf81e1af4f7d9101d47a6a32505ed2a5ae2c13a7))



# [3.2.0](https://github.com/SidebarJS/sidebarjs/compare/3.1.0...3.2.0) (2017-11-16)


### Features

* **definition:** no more version prop in SidebarElement ([f515fcc](https://github.com/SidebarJS/sidebarjs/commit/f515fccd1983802632f8980c072ec5b6a5c295ff))



# [3.1.0](https://github.com/SidebarJS/sidebarjs/compare/3.0.0...3.1.0) (2017-09-04)


### Bug Fixes

* **ts:** create method in definition file ([dae2869](https://github.com/SidebarJS/sidebarjs/commit/dae28696877589fe8e4aa7be71e613c230d8ec8d))



# [3.0.0](https://github.com/SidebarJS/sidebarjs/compare/2.2.0...3.0.0) (2017-08-08)


### Bug Fixes


src/sidebarjs.ts
import { HTMLSidebarElement, SidebarBase, SidebarConfig } from '../index';

const sidebarjs = 'sidebarjs';
const isVisible = `${sidebarjs}--is-visible`;
const isMoving = `${sidebarjs}--is-moving`;
const LEFT_POSITION = 'left';
const RIGHT_POSITION = 'right';
const TRANSITION_DURATION = 400;
const POSITIONS = [LEFT_POSITION, RIGHT_POSITION];

export default class SidebarJS implements SidebarBase {
  public component: HTMLElement;
  public container: HTMLElement;
  public background: HTMLElement;
  public documentMinSwipeX: number;
  public documentSwipeRange: number;
  public nativeSwipe: boolean;
  public nativeSwipeOpen: boolean;
  public position: string;
  private initialTouch: number;
  private touchMoveSidebar: number;
  private openMovement: number;

  constructor({
                component,
                container,
                background,
                documentMinSwipeX,
                documentSwipeRange,
                nativeSwipe,
                nativeSwipeOpen,
                position,
              }: SidebarConfig = {}) {
    this.component = component || <HTMLElement> document.querySelector(`[${sidebarjs}]`);
    this.container = container || SidebarJS.create(`${sidebarjs}-container`);
    this.background = background || SidebarJS.create(`${sidebarjs}-background`);
    this.documentMinSwipeX = documentMinSwipeX || 10;
    this.documentSwipeRange = documentSwipeRange || 40;
    this.nativeSwipe = nativeSwipe !== false;
    this.nativeSwipeOpen = nativeSwipeOpen !== false;

    const hasAllConfigDOMElements = component && container && background;
    if (!hasAllConfigDOMElements) {
      try {
        this.transcludeContent();
      } catch (e) {
        throw new Error('You must define an element with [sidebarjs] attribute');
      }
    }

    if (this.nativeSwipe) {
      this.addNativeGestures();
      if (this.nativeSwipeOpen) {
        this.addNativeOpenGestures();
      }
    }

    this.setPosition(position);
    this.addAttrsEventsListeners(this.component.getAttribute(sidebarjs));
    this.background.addEventListener('click', this.close.bind(this));
  }

  public toggle(): void {
    this.component.classList.contains(isVisible) ? this.close() : this.open();
  }

  public open(): void {
    this.component.classList.add(isVisible);
  }

  public close(): void {
    this.component.classList.remove(isVisible);
  }

  public isVisible(): boolean {
    return this.component.classList.contains(isVisible);
  }

  public setPosition(position: string): void {
    this.component.classList.add(isMoving);
    this.position = POSITIONS.indexOf(position) >= 0 ? position : LEFT_POSITION;
    POSITIONS.forEach((POS) => this.component.classList.remove(`${sidebarjs}--${POS}`));
    this.component.classList.add(`${sidebarjs}--${this.hasRightPosition() ? RIGHT_POSITION : LEFT_POSITION}`);
    setTimeout(() => this.component.classList.remove(isMoving), TRANSITION_DURATION);
  }

  public addAttrsEventsListeners(sidebarName: string): void {
    const actions = ['toggle', 'open', 'close'];
    for (let i = 0; i < actions.length; i++) {
      const elements = document.querySelectorAll(`[${sidebarjs}-${actions[i]}="${sidebarName}"]`);
      for (let j = 0; j < elements.length; j++) {
        if (!SidebarJS.elemHasListener(<HTMLElement> elements[j])) {
          elements[j].addEventListener('click', this[actions[i]].bind(this));
          SidebarJS.elemHasListener(<HTMLElement> elements[j], true);
        }
      }
    }
  }

  private hasLeftPosition(): boolean {
    return this.position === LEFT_POSITION;
  }

  private hasRightPosition(): boolean {
    return this.position === RIGHT_POSITION;
  }

  private transcludeContent(): void {
    this.container.innerHTML = this.component.innerHTML;
    this.component.innerHTML = '';
    this.component.appendChild(this.container);
    this.component.appendChild(this.background);
  }

  private addNativeGestures(): void {
    this.component.addEventListener('touchstart', this.onTouchStart.bind(this));
    this.component.addEventListener('touchmove', this.onTouchMove.bind(this));
    this.component.addEventListener('touchend', this.onTouchEnd.bind(this));
  }

  private addNativeOpenGestures(): void {
    document.addEventListener('touchstart', this.onSwipeOpenStart.bind(this));
    document.addEventListener('touchmove', this.onSwipeOpenMove.bind(this));
    document.addEventListener('touchend', this.onSwipeOpenEnd.bind(this));
  }

  private onTouchStart(e: TouchEvent): void {
    this.initialTouch = e.touches[0].pageX;
  }

  private onTouchMove(e: TouchEvent): void {
    const documentSwiped = this.initialTouch - e.touches[0].clientX;
    const sidebarMovement = this.getSidebarPosition(documentSwiped);
    this.touchMoveSidebar = -documentSwiped;
    if (sidebarMovement <= this.container.clientWidth) {
      this.moveSidebar(this.touchMoveSidebar);
    }
  }

  private onTouchEnd(): void {
    this.component.classList.remove(isMoving);
    Math.abs(this.touchMoveSidebar) > (this.container.clientWidth / 3.5) ? this.close() : this.open();
    this.container.removeAttribute('style');
    this.background.removeAttribute('style');
    delete this.initialTouch;
    delete this.touchMoveSidebar;
  }

  private moveSidebar(movement: number): void {
    this.component.classList.add(isMoving);
    SidebarJS.vendorify(this.container, 'transform', `translate(${movement}px, 0)`);
    this.changeBackgroundOpacity(movement);
  }

  private changeBackgroundOpacity(movement: number): void {
    const opacity = 0.3 - (Math.abs(movement) / (this.container.clientWidth * 3.5));
    this.background.style.opacity = (opacity).toString();
  }

  private onSwipeOpenStart(e: TouchEvent): void {
    if (this.targetElementIsBackground(e)) {
      return;
    }
    const {clientWidth} = document.body;
    const touchPositionX = e.touches[0].clientX;
    const documentTouch = this.hasLeftPosition() ? touchPositionX : clientWidth - touchPositionX;
    if (documentTouch < this.documentSwipeRange) {
      this.onTouchStart(e);
    }
  }

  private onSwipeOpenMove(e: TouchEvent): void {
    if (!this.targetElementIsBackground(e) && this.initialTouch && !this.isVisible()) {
      const documentSwiped = e.touches[0].clientX - this.initialTouch;
      const sidebarMovement = this.getSidebarPosition(documentSwiped);
      if (sidebarMovement > 0) {
        SidebarJS.vendorify(this.component, 'transform', 'translate(0, 0)');
        SidebarJS.vendorify(this.component, 'transition', 'none');
        this.openMovement = sidebarMovement * (this.hasLeftPosition() ? -1 : 1);
        this.moveSidebar(this.openMovement);
      }
    }
  }

  private onSwipeOpenEnd(): void {
    if (this.openMovement) {
      delete this.openMovement;
      this.component.removeAttribute('style');
      this.onTouchEnd();
    }
  }

  private getSidebarPosition(swiped: number): number {
    return (this.container.clientWidth - (this.hasLeftPosition() ? swiped : -swiped));
  }

  private targetElementIsBackground(e: TouchEvent): boolean {
    const touchedElement = <HTMLElement> e.target;
    return touchedElement.hasAttribute(`${sidebarjs}-background`);
  }

  public static create(element: string): HTMLElement {
    const el = document.createElement('div');
    el.setAttribute(element, '');
    return el;
  }

  public static vendorify(el: HTMLElement, prop: string, val: string): HTMLElement {
    const Prop = prop.charAt(0).toUpperCase() + prop.slice(1);
    const prefs = ['Moz', 'Webkit', 'O', 'ms'];
    el.style[prop] = val;
    for (let i = 0; i < prefs.length; i++) {
      el.style[prefs[i] + Prop] = val;
    }
    return el;
  }

  public static elemHasListener(elem: HTMLSidebarElement, value?: boolean): boolean {
    return elem && (value === true || value === false) ? elem.sidebarjsListener = value : !!elem.sidebarjsListener;
  }

  static get version(): string {
    return '2.2.0';
  }
}
